fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,431 @@
+| `confirmation_byzantine_threshold` | `uint64` | `33`                               | the maximum percentage of Byzantine validators among the validator set |
+| `confirmation_slashing_threshold`  | `uint64` | `confirmation_byzantine_threshold` | the maximum percentage of slashings among the validator set            |
+
+
+### Helper Functions
+
+#### `get_committee_weight_between_slots`
+
+```python
+def get_committee_weight_between_slots(store: Store, start_slot: Slot, end_slot: Slot) -> Gwei:
+    """Returns the total weight of committees between ``start_slot`` and ``end_slot`` (inclusive of both).
+    Uses the justified state to compute committee weights.
+    """
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    total_active_balance = get_total_active_balance(justified_state)
+
+    # If an entire epoch is covered by the range, return the total active balance
+    start_epoch = compute_epoch_at_slot(start_slot)
+    end_epoch = compute_epoch_at_slot(end_slot)
+    if end_epoch > start_epoch + 1:
+        return total_active_balance
+
+    committee_weight = total_active_balance // SLOTS_PER_EPOCH
+    if start_epoch == end_epoch:
+        num_committees = end_slot - start_slot + 1
+    else:
+        # A range that spans an epoch boundary, but does not span any full epoch
+        # needs pro-rata calculation
+        # First, calculate the number of committees in the current epoch
+        num_slots_in_current_epoch = (end_slot % SLOTS_PER_EPOCH) + 1
+        # Next, calculate the number of slots remaining in the current epoch
+        remaining_slots_in_current_epoch = SLOTS_PER_EPOCH - num_slots_in_current_epoch
+        # Then, calculate the number of slots in the previous epoch
+        num_slots_in_previous_epoch = SLOTS_PER_EPOCH - (start_slot % SLOTS_PER_EPOCH)
+
+        # Each committee from the previous epoch only contributes a pro-rated weight
+        multiplier = remaining_slots_in_current_epoch / SLOTS_PER_EPOCH
+        num_committees = num_slots_in_current_epoch + num_slots_in_previous_epoch * multiplier
+    return num_committees * committee_weight
+```
+
+#### `is_one_confirmed`
+
+```python
+def is_one_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    parent_block = store.blocks[block.parent_root]
+    support = int(get_weight(store, block_root))
+    maximum_support = int(get_committee_weight_between_slots(store, Slot(parent_block.slot + 1), current_slot))
+    proposer_score = int(get_proposer_score(store))
+
+    return (
+        support / maximum_support >
+        0.5 * (1 + proposer_score / maximum_support) + confirmation_byzantine_threshold / 100
+    )
+```
+
+#### `is_lmd_confirmed`
+
+```python
+def is_lmd_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    if block_root == store.finalized_checkpoint.root:
```

Comment:
We should look into this.

Reply:
I think that justified block is definitely lmd-confirmed under the confirmation rule assumptions. So, it makes sense to me to switch to justified root here. I think this function can be simplified in the following way:

```python
def is_lmd_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
    if block_root == store.justified_checkpoint.root:
        return True

    if not is_one_confirmed(store, confirmation_byzantine_threshold, block_root):
        return False

    return is_lmd_confirmed(store, confirmation_byzantine_threshold, block.parent_root)
```

with the finalized ancestor check moved to `is_confirmed`:
```python
# Check block is a descendant of the finalized block at the checkpoint finalized slot
finalized_checkpoint_block = get_checkpoint_block(store, block_root, store.finalized_checkpoint.epoch)
assert store.finalized_checkpoint.root == finalized_checkpoint_block
```

Reply:
Firstly, I think we can move the "finalized descendant" check to `is_confirmed` method to be called only once.

Secondly, I think that `store.justified_checkpoint` is safe to be used here though it is not necessarily LMD confirmed. Honest validators respect `store.justified_checkpoint` in the FC filtering and can never vote for any chain not containing it, if we assume synchrony then honest validators in current epoch can't justify a checkpoint conflicting to `store.justified_checkpoint`. I am fine to traverse back to finalized checkpoint unless there is a proof that this optimisation is safe.

Reply:
> I think we can move the "finalized descendant" check to is_confirmed method to be called only once.

Just realized that I missed the following piece of information in my reply.

My proposal was to add the following check in `is_confirmed`
```python
def is_confirmed(...):
    ....
    finalized_checkpoint_block = get_checkpoint_block(store, block_root, store.finalized_checkpoint.epoch)
    if store.finalized_checkpoint.root != finalized_checkpoint_block:
        return False
   ...
```
and for the assert at the beginning of the method `is_lmd_confirmed` to play the role of method precondition.

Not sure though whether this style is in line with the rest of the specification.

Reply:
To be able to confirm the latest justified checkpoint, so that we don't have to go all the way back to finalized, we just need to check whether or not it is possible for a higher checkpoint to be justified. For example, we could check whether:
- Epoch of the justified checkpoint is current epoch - 1
- Whether it is possible for a conflicting one to be justified in this epoch (similar to what `is_ffg_confirmed` does)

Reply:
I have just implemented the early exit criterion that @fradamt suggested [here](https://github.com/ethereum/consensus-specs/pull/3339#discussion_r1266703220).